00-basics/002-javascript/freeCodeCamp_03_22.js

```diff
@@ -0,0 +1,402 @@
+let contacts = [
+    {
+        "firstName": "Akira",
+        "lastName": "Laine",
+        "number": "0543236543",
+        "likes": ["Pizza", "Coding", "Brownie Points"]
+    },
+    {
+        "firstName": "Harry",
+        "lastName": "Potter",
+        "number": "0994372684",
+        "likes": ["Hogwarts", "Magic", "Hagrid"]
+    },
+    {
+        "firstName": "Sherlock",
+        "lastName": "Holmes",
+        "number": "0487345643",
+        "likes": ["Intriguing Cases", "Violin"]
+    },
+    {
+        "firstName": "Kristian",
+        "lastName": "Vos",
+        "number": "unknown",
+        "likes": ["JavasScript", "Gaming", "Foxes"]
+    }
+];
+
+function lookUpProfile(name, prop) {
+    for (let i = 0; i < contacts.length; i ++) {
+       if (contacts[i].firstName === name) {
+    return contacts[i][prop] || "No such property";
+       }
+    }
+    return "No such contact"
+}
+console.log(lookUpProfile("Harry", "lastName"));
+
+function randomFraction() {
+    return Math.random();
+}
+console.log((randomFraction()));
+
+function randomWholeNum() {
+    return Math.floor(Math.random() * 10);
+}
+console.log(randomWholeNum());
+
+function randomRange(myMin, myMax) {
+    return Math.floor(Math.random() * (myMax - myMin + 1)) + myMin;
+}
+let myRandom = randomRange(9, 15);
+
+console.log(myRandom);
+
+function convertTooInteger(str) {
+    return parseInt(str);
+}
+convertTooInteger("56");
+console.log(convertTooInteger("56"));
+
+function convertToInteger(str) {
+    return parseInt(str, 2);
+}
+convertToInteger("10011");
+console.log(convertToInteger("10011"));
+
+function checkEqual(a, b) {
+    return a === b ? "Equal" : "Not Equal";
+}
+
+console.log(checkEqual(1, 2));
+
+function checkSign(num) {
+    return  (num > 0) ? "positive" : (num < 0) ? "negative" : "zero";
+}
+console.log(checkSign(10));
+
+function countDown (n) {
+    if (n < 1) {
+        return [];
+    } else {
+        const arr = countDown(n - 1);
+        arr.unshift(n);
+        return arr;
+    }
+}
+console.log(countDown(5));
+
+function rangeOfNumbers(startNum, endNum) {
+    if (endNum - startNum === 0) {
+        return [startNum];
+    } else {
+        let numbers = rangeOfNumbers(startNum, endNum - 1);
+        numbers.push(endNum);
+        return numbers;
+    }
+}
+console.log(rangeOfNumbers(3, 5));
+
+let catName ;
+let quote ;
+function catTalk() {
+    "use strict";
+    catName = "Oliver ";
+    quote = catName + "says Meow!";
+
+}
+catTalk();
+console.log(quote);
+// How to understand "Variable might not have been initialized" (quote) ?
```

Comment:
`let quote` declares a variable, but does not yet set it to any value.
Value is only assigned later, inside `catTalk()`. Now I think that IDE is being overly paranoid here, because technically variable definitely will be set after calling the `catTalk()`, there are no branches or conditions that would lead `quote` not to be set there. This warning is valid in this case:
```
function someFunction(a)
let x
if (a === 1) {
x = 'text'
}
console.log(x)
```

In this example `x` will not be initialized in case `a` is not equal to `1`.